### DISCUSSION TITLE
Sub app does not include sub app root path in OpenAPI endpoint paths

### DISCUSSION THREAD

**Discussion by totalhack** [fastapi/fastapi#9404] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# NOTE: this is a simple illustration of the problem, not a full reproduction

from fastapi import FastAPI

# Main app

app = FastAPI(
    title="Test"
    # Open API spec for main app, no issues here
    openapi_url="/openapi.json",
    docs_url=None,
    redoc_url="/docs",
)

# Adding a sub-app

plugin_app = FastAPI(
    title="Plugin API",
    description="XYZ",
    version="0.0.1",
    servers=[{"url": f"{settings.SERVER_HOST}"}],
    root_path_in_servers=False,
    openapi_url="/openapi.json",
    docs_url=None,
    redoc_url="/docs"
)

# Mount it at /plugin

app.mount("/plugin", plugin_app)

# Here's the issue...
# This creates a route at /plugin/foo as expected but /plugin/openapi.json 
# lists the path as just /foo which is wrong!

@plugin_app.post("/foo")
def foo(request:
    return {"a": "b"}
```

### Description

The auto-generated OpenAPI spec at plugin/openapi.json does not include /plugin as a prefix to its paths, so those paths are all invalid. 

### Operating System

Linux

### Operating System Details

Ubuntu 22.04

### FastAPI Version

0.95.1

### Python Version

3.9.16

**Comment by hugohutri** [comment]
We faced the same problem, and ended up removing the sub apps because of this and using only a one root app.

FastApi version: 0.110.3
Python version: 3.10

**Comment by YuriiMotov** [comment]
I don't think something is wrong here.

FastAPI generates docs for sub-app as if it's standalone app. And by default FastAPI puts mounting path (`/plugin`) to servers. But you explicitly disabled this (`servers=[{"url": f"{settings.SERVER_HOST}"}]`, and `root_path_in_servers=False`).
If you remove those parameters and use defaults, everything looks good.

Am I missing something?

**Reply by PGijsbers** [reply]
I just created a fresh environment (py 3.11.3, fastapi 0.116.1) and used the example code exactly as provided in [Sub Applications](https://fastapi.tiangolo.com/advanced/sub-applications/#check-the-automatic-api-docs) documentation and run it as instructed. When I navigate to `http://localhost/subapi/docs` I see the same behavior: the root path is added to the server list and the paths displays in the documentation are relative. I.e., I see `/sub` and not `/subapi/sub`, as well as a "servers" dropdown at the top. The REST API itself functions as expected. However, FastAPI's Sub Applications documentation shows a picture that the swagger docs should render the endpoint as `subapi/sub`, which is how I want it. Am I missing something?

**Reply by YuriiMotov** [reply]
You a right that picture in docs doesn't reflect the current situation. Seems that it needs to be updated
<img width="674" height="274" alt="image" src="https://github.com/user-attachments/assets/3127515a-5f13-4f91-a385-b271b0650247" />

If you need `root_path` to be included to paths, the only way I know is to modify the openapi schema using approach described here: https://fastapi.tiangolo.com/how-to/extending-openapi

**Reply by PGijsbers** [reply]
For those looking for an example on how to do this, this is a minimal example to get you started (I'm not sure yet if there are other aspects I should take into account):
```python
from fastapi import FastAPI

app = FastAPI()


@app.get("/app")
def read_main():
    return {"message": "Hello World from main app"}


subapi = FastAPI()

def custom_api():
    if subapi.openapi_schema:
        return subapi.openapi_schema
    schema = subapi._openapi()
    subapi.openapi_schema = schema

    # By including the prefix in the path, we need to adjust the automatically added server list.
    # If you want the dropdown, manually set `'servers'`, otherwise you can just delete it here 
    # (or set root_path_in_servers when creating the subapp).
    # schema['servers'] = [{"url":"/"}]
    del schema['servers']
    
    paths = schema['paths'].copy()
    for path, metadata  in paths.items():
      schema['paths'][f"/subapi{path}"] = metadata
      del schema['paths'][path]
    return schema

subapi._openapi = subapi.openapi
subapi.openapi = custom_api


@subapi.get("/sub")
def read_sub():
    return {"message": "Hello World from sub API"}


app.mount("/subapi", subapi)
```

The server and docs are both functional and the docs are rendered as:
<img width="1489" height="937" alt="afbeelding" src="https://github.com/user-attachments/assets/ace19076-6f39-4f45-b46e-2dd0e11c246d" />

**Reply by ivica3730k** [reply]
Hi all! Thank you very much for your input and code samples.

I attempted something similar, by doing this:

```python3
def custom_api_spec(app: FastAPI) -> dict:

    fastapi_path_prefix = app.root_path

    if app.openapi_schema:
        return app.openapi_schema
    schema = app.openapi()
    app.openapi_schema = schema

    schema.pop("servers", None)  # Remove servers to avoid confusion, we dont point docs to specific server urls

    paths = schema["paths"].copy()
    for path, metadata in paths.items():
        schema["paths"][f"{fastapi_path_prefix}{path}"] = metadata
        del schema["paths"][path]
    return schema
```

and then I make app using this:

```python3
def _create_fastapi_app(root_url: str) -> FastAPI:
    return FastAPI(
        title=f"{settings.service_name} API",
        description="Backend FastAPI project",
        version="0.1.0",
        docs_url=None,
        openapi_url=None,
        redoc_url=None,
        root_path=root_url,
    )
 ```
 
 and use all of above to create the documentation router:
 ```python3
 def get_docs_router(app: FastAPI, doc_url_prefix: str) -> APIRouter:
    """
    Returns an APIRouter with Swagger, Redoc, and OpenAPI endpoints for a specific FastAPI app.
    """
    router = APIRouter()

    @router.get("/", include_in_schema=False)
    async def root_docs(request: Request):
        html_content = f"""
        <html>
            <head>
                <title>API Documentation</title>
            </head>
            <body>
                <h1>API Documentation</h1>
                <p>Navigate to <a href="{doc_url_prefix}/swagger">Swagger UI</a> or <a href="{doc_url_prefix}/redocly">Redoc documentation</a>.</p>
            </body>
        </html>
        """
        return HTMLResponse(content=html_content, status_code=200)

    @router.get("/redocly", include_in_schema=False)
    async def redoc_docs(request: Request):
        return get_redoc_html(openapi_url=f"{doc_url_prefix}/openapi.json", title=f"{app.title} Docs")

    @router.get("/swagger", include_in_schema=False)
    async def swagger_docs(request: Request):
        return get_swagger_ui_html(openapi_url=f"{doc_url_prefix}/openapi.json", title=f"{app.title} Docs")

    @router.get("/openapi.json", include_in_schema=False)
    async def openapi_json(request: Request):
        return custom_api_spec(app)

    return router
```

im curious is this something that a lot of people might want to do? **if yes - would it worth it to have it configurable on the FastAPI() init fn?**
